src/main/java/org/truffleruby/core/regexp/TruffleRegexpNodes.java

```diff
@@ -108,4 +136,281 @@ protected int getCacheLimit() {
         }
     }
 
+    @CoreMethod(names = "compilation_stats_array", onSingleton = true, required = 0)
+    public static abstract class CompilationStatsArrayNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public Object buildStatsArray(@Cached("create()") ArrayBuilderNode arrayBuilderNode) {
+            Object store = arrayBuilderNode.start(compiledRegexps.size() * 2);
+            int n = 0;
+            for (Entry<RegexpInfo, AtomicInteger> e : compiledRegexps.entrySet()) {
+                Rope key = StringOperations.encodeRope(e.getKey().toString(), UTF8Encoding.INSTANCE);
+                store = arrayBuilderNode.appendValue(store, n++, StringOperations.createString(getContext(), key));
+                store = arrayBuilderNode.appendValue(store, n++, e.getValue().get());
+            }
+
+            return createArray(store, n);
+        }
+    }
+
+    @CoreMethod(names = "match_stats_array", onSingleton = true, required = 0)
+    public static abstract class MatchStatsArrayNode extends CoreMethodArrayArgumentsNode {
+
+        @TruffleBoundary
+        @Specialization
+        public Object buildStatsArray(@Cached("create()") ArrayBuilderNode arrayBuilderNode) {
+            Object store = arrayBuilderNode.start(matchedRegexps.size() * 2);
+            int n = 0;
+            for (Entry<MatchInfo, AtomicInteger> e : matchedRegexps.entrySet()) {
+                Rope key = StringOperations.encodeRope(e.getKey().toString(), UTF8Encoding.INSTANCE);
+                store = arrayBuilderNode.appendValue(store, n++, StringOperations.createString(getContext(), key));
+                store = arrayBuilderNode.appendValue(store, n++, e.getValue().get());
+            }
+
+            return createArray(store, n);
+        }
```

Comment:
Looks like these 2 could share a method or even be a single node, taking some extra argument to decide what information should be used.